lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,538 @@
+                .map_or(usize::max_value(), NonZeroUsize::get),
+            batches: HashMap::default(),
+            closed_batches: Vec::default(),
+            timer,
+            partitioner,
+            stream,
+        }
+    }
+}
+
+impl<St, Prt, T> Stream for Batcher<St, Prt, T>
+where
+    St: Stream<Item = Prt::Item> + Unpin,
+    Prt: Partitioner + Unpin,
+    Prt::Item: ByteSizeOf,
+    T: Timer,
+{
+    type Item = (Option<Prt::Key>, Vec<Prt::Item>);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            if !this.closed_batches.is_empty() {
+                return Poll::Ready(this.closed_batches.pop());
+            }
+            match this.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    if this.timer.has_elapsed() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        this.timer.reset();
+                        continue;
+                    }
+                    return Poll::Pending;
+                }
+                Poll::Ready(None) => {
+                    // Now that the underlying stream is closed we need to clear
+                    // out our batches.
+                    if !this.batches.is_empty() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        continue;
+                    }
+                    return Poll::Ready(None);
+                }
+                Poll::Ready(Some(item)) => {
+                    let item_key = this.partitioner.partition(&item);
+
+                    if let Some(batch) = this.batches.get_mut(&item_key) {
+                        if batch.has_space(&item) {
+                            batch.push(item);
+                        } else {
+                            let new_batch =
+                                Batch::new(*this.batch_item_limit, *this.batch_allocation_limit)
+                                    .with(item);
+                            let batch = mem::replace(batch, new_batch);
+                            let (_, arr) = batch.destruct();
+                            if this.timer.has_elapsed() {
+                                this.closed_batches.extend(
+                                    this.batches
+                                        .drain()
+                                        .map(|(key, batch)| (key, batch.destruct().1)),
+                                );
```

Comment:
I could be wrong, but it looks like this might be closing the new batch immediately with just the single item.

The `mem::replace` effectively inserts `new_batch` into `this.batches`, which we then `drain()` unconditionally into `this.closed_batches`.

Reply:
Oh, nope, timer check makes this mostly fine I think. Not unconditional at all.

Reply:
Oh yeah, that is exactly what's happening. Nice catch.

Reply:
Resolved in 3dccc44